hoomd/ParticleGroup.cc

```diff
@@ -789,6 +792,100 @@ unsigned int ParticleGroup::intersectionSize(std::shared_ptr<ParticleGroup> othe
     return n;
     }
 
+void ParticleGroup::thermalizeParticleMomenta(Scalar kT, unsigned int seed, unsigned int timestep)
+    {
+    unsigned int group_size = this->getNumMembers();
+
+    const unsigned int D = Scalar(m_sysdef->getNDimensions());
+
+    ArrayHandle<Scalar4> h_vel(m_pdata->getVelocities(),
+                               access_location::host,
+                               access_mode::readwrite);
+
+    ArrayHandle<Scalar4> h_orientation(m_pdata->getOrientationArray(),
+                                       access_location::host,
+                                       access_mode::read);
+
+    ArrayHandle<Scalar4> h_angmom(m_pdata->getAngularMomentumArray(),
+                                  access_location::host,
+                                  access_mode::readwrite);
+
+    ArrayHandle<Scalar3> h_inertia(m_pdata->getMomentsOfInertiaArray(),
+                                   access_location::host,
+                                   access_mode::read);
+
+    ArrayHandle<unsigned int> h_tag(m_pdata->getTags(),
+                                    access_location::host,
+                                    access_mode::read);
+
+    // Total the system's linear momentum
+    vec3<Scalar> tot_momentum(0, 0, 0);
+
+    // Loop over all particles in the group
+    for (unsigned int group_idx = 0; group_idx < group_size; group_idx++)
+        {
+        unsigned int j = this->getMemberIndex(group_idx);
+        unsigned int ptag = h_tag.data[j];
+
+        // Seed the RNG
+        hoomd::RandomGenerator rng(hoomd::RNGIdentifier::ParticleGroupThermalize, seed, ptag, timestep);
```

Comment:
Any reason not to just use index `j` directly in the rng and avoid accessing the tags? Saves having to map the particle tags array unless there's some reason you need the consistency of a tag rather than the index for reproducibility reasons.

Reply:
This method acts locally on each rank. To prevent the same random numbers from being used we either need to use tags or include the MPI rank as a seed. Using the tags is consistent with how BD and DPD seed their RNGs, so that was propagated here. I don't think one additional memory lookup is going to hurt performance much in a rarely called method. Plus, as you say it does provide some level of consistency across serial and MPI runs.

Reply:
Agreed, no need to optimize here really. The consistency is good.